### ISSUE TITLE
Efficient top list for metrics data source

### ISSUE THREAD

**Issue by srikanthccv** [SigNoz/signoz#6204]
Today, the query fetches all the data, caches it and does the in-process calculation to apply the limit or not. While it is best for many cases, it is not the right tool where there can be several thousand series in the result. Implement better top-list support.

**Comment by namanv3** [comment]
@srikanthccv this issue is about not having order by and limit application in this file right?
https://github.com/SigNoz/signoz/blob/main/pkg/types/querybuildertypes/querybuildertypesv5/series_limit.go
And instead to include it in the clickhouse query built by `statement_builder.go` correct?

**Comment by srikanthccv** [comment]
Yes, that's correct. We don't need to fetch tens of thousands of pods usage into application memory if user is just interested in top 10 pods.

**Comment by srikanthccv** [comment]
Some context why we do this. For metric queries, regardless of the limit, we end up reading all the data, since we do that, we went bit greedy and cache all of that. 1. What do we gain by moving the limit to clickhouse? 2. and what do we lose 3. How will formulas work?

**Comment by namanv3** [comment]
But if the cache fingerprint contains the limit, then how does the cache even help. If the first time with limit 10, the cache has the 1000 rows, and then the user searches for top 20, it won't even hit the cache.

**Comment by srikanthccv** [comment]
You are right, the limit wasn't part of the metrics queries in the initial implementation plan but I see that it has been part of fingerprint as long as the qbv5 existed so the assumed caching benefit never actually helped.

**Comment by namanv3** [comment]
@srikanthccv I have the change ready, but I was wondering if we should instead fix the fingerprint issue and make the cache actually work. If a user has put a limit, they are more likely to change it to other values, and that's where the cached data will really help.

**Comment by srikanthccv** [comment]
> If a user has put a limit, they are more likely to change it to other values, and that's where the cached data will really help.

That was the reason v4/query_range doesn't add the limit to cache fingerprint. However, we missed it for v5, so it's neither helping nor resource efficient.
 https://github.com/SigNoz/signoz/blob/48ccdfcf64b743757ebbb02fc9bb145309beaaf9/pkg/query-service/app/queryBuilder/query_builder.go#L400-L438
The main problem with applying series limit at post process is the cache size, we have a limited cache given for application cache, if we bring all series and put it into cache when the user is just interested in very few series, the cache size grows and evicts the other important cached data leading to inefficient user of the cache system. The past reasoning that we were anyway reading all data in ClickHouse why not cache the entire thing so we can speed up the next set of queries didn't take other things into account, specifically cache size (we had unbound cache earlier but it's bounded now). So given this constraint, does it actually make sense to fetch all series and apply limit in post process? 

Before I look at the PR change
- please share the ClickHouse query and write how does it solve the issue?
- how do we deal with formula queries?
-  does it actually make sense to fetch all series and apply limit in post process?

**Comment by namanv3** [comment]
@srikanthccv here's the doc: https://www.notion.so/signoz/Limits-and-sorts-319fcc6bcd19806e8212c7d8916b15d0?source=copy_link
Also, the PR is not correct right now, I'll mark it ready for review once it is

**Comment by namanv3** [comment]
Closed the linked PR because of an edge case that's described here: https://www.notion.so/signoz/Limit-in-CH-breaks-because-of-cache-32dfcc6bcd1980a99b37f1cbe3a4b9c2?source=copy_link
Need to do some additional research on how to solve this.
